README.md
# rw_tcx


name: rw_tcx

description: A simple package to read and write TCX file using Dart
With this package it is possible to extract track points for a TCX file or write a valid TCX file from an array of track points 


## Getting Started


It is a Dart only package, Flutter is not needed

 


## How to install
Check on pub.dev/packages to see how to install this package
https://pub.dev/packages/rw_tcx#-installing-tab-


## How to use it

There are only 2 APIs
- readTCX
- writeTCX

## Example

There is a complete example showing how to use readTCX and writeTCX
For readTCX, the sample2.tcx stored in assets is coming from Garmin Connect export
For writeTCX, the generated TCX file is sent to Google Drive in addition to be uploaded to Strava


## To check your TCX file

There is a nice online tool https://www.gpsies.com/convert.do to get a detailed explanation of what is wrong in your TCX file 



## Tested on:

So far only on Android 8.0 for the moment

## Contributors welcome!

If you spot a problem/bug or if you consider that the code could be better please post a new issue.
I have developed this package to be able to upload a TCX to Strava so rTCX.dart is very basic. 
Feel free to contribute to add more features.





## License:
rw_tcx is provided under a MIT License. Copyright (c) 2019 Patrick FINKELSTEIN




CHANGELOG.md
## [0.0.1] - 28/08/2019

* Initial release

## [0.0.2] - 29/08/2019  

* Added google drive export to example


## [0.1.0] - 29/08/2019

* Initial working release

## [0.1.0+2] - 29/08/2019

* Longer description in pubspec.yaml

test/rw_tcx_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:rw_tcx/rTCX.dart';
import 'package:rw_tcx/wTCX.dart';
import 'package:rw_tcx/models/TCXModel.dart';

void main() {
  group('readTCX test', () {


    test('SearchElement tests', () {
    

    String contents = "<TotalTimeSeconds>16524.0</TotalTimeSeconds>";
    Tag expectedTag = Tag(content: '16524.0', index: 25);
    Tag resultTag = searchElement('TotalTimeSeconds', contents);
    expect(resultTag.content, equals(expectedTag.content));

    Tag wrongResultTag =searchElement('wrong', contents);
    Tag emptyTag =Tag(content: '', index:0);
    expect(wrongResultTag.content, equals(emptyTag.content));
    

    });


    test('SearchAttribute tests', () {

      Tag expectedTag = Tag(content: ' Sport="Biking"', index: 34);
      Tag resultTag = searchAttribute('Activity', ' hjhkjhkj <Activity Sport="Biking"> hjhkjhjk');
      expect(expectedTag.content, equals(resultTag.content));
      expect(expectedTag.index, equals(resultTag.index));
    
    });


    test('SearchExtension tests', () {
      String textExtension = '''<AltitudeMeters>79.80000305175781</AltitudeMeters>
            <DistanceMeters>2.75</DistanceMeters>
            <Extensions>
              <ns3:TPX>
                <ns3:Speed>2.7809998989105225</ns3:Speed>
              </ns3:TPX>
            </Extensions>
          </Trackpoint>''';
      
      String expectedResult = '2.7809998989105225';
      String result = searchExtension('Speed', textExtension);

      expect(expectedResult, equals(result));
    });

    
  });



  group('write TCX test', () {

    test('CreateTimestamp tests', () {
    var date = DateTime(2019, 6, 11, 16, 50);
    String resultTimestamp = createTimestamp(date);
   
    expect(resultTimestamp, equals('2019-06-11T14:50:00.000Z'));


   

    });

    test('addExtension tests', ()  {

      String resultExtension = '';
      String extensionContent = """<Extensions>\n   <ns3:TPX>\n     <ns3:Speed>1.996999979019165</ns3:Speed>\n   </ns3:TPX>\n</Extensions>\n""";
      resultExtension = addExtension('Speed', 1.996999979019165);
      expect(resultExtension, equals(extensionContent));
    });



  test('addHeartRate tests', ()  {

      String resultExtension = '';
      String heartRateContent = """
                 <HeartRateBpm>
              <Value>61</Value>
            </HeartRateBpm>\n""";

      resultExtension = addHeartRate(61);
      expect(resultExtension, equals(heartRateContent));
    });

  });
}


lib/wTCX.dart
// wTCX.dart
// Tools to generate a TCX file
// To test on Strava

import 'dart:io';
import 'package:path_provider/path_provider.dart';

import 'models/TCXModel.dart';
import 'logTool.dart';

// Generate the TCX file
/// from a TCX Model
/// Store the TCX file in genera
/// TODO: Add return code
///
// Future<String> writeTCX(TCXModel tcxInfos, String filename) async {
Future<void> writeTCX(TCXModel tcxInfos, String filename) async {
  Future<File> _localFile(String fileName) async {
    final directory = await getApplicationDocumentsDirectory();
    var path = directory.path;
    return File('$path/$fileName');
  }

  // Now generate a new file from rideData
  var generatedTCXFile = await _localFile(filename);
  var sink = generatedTCXFile.openWrite(mode: FileMode.writeOnly);

  String contents = '';

  // Generate the prolog of the TCX file
  final String prolog = """ <?xml version="1.0" encoding="UTF-8"?>
    <TrainingCenterDatabase
    xsi:schemaLocation="http://www.garmin.com/xmlschemas/TrainingCenterDatabase/v2 http://www.garmin.com/xmlschemas/TrainingCenterDatabasev2.xsd"
    xmlns:ns5="http://www.garmin.com/xmlschemas/ActivityGoals/v1"
    xmlns:ns3="http://www.garmin.com/xmlschemas/ActivityExtension/v2"
    xmlns:ns2="http://www.garmin.com/xmlschemas/UserProfile/v2"
    xmlns="http://www.garmin.com/xmlschemas/TrainingCenterDatabase/v2"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:ns4="http://www.garmin.com/xmlschemas/ProfileExtension/v1">\n""";

  final String tailActivity = """      <Creator xsi:type="Device_t">
        <Name>rw_TCX</Name>
        <UnitId>3873921795</UnitId>
        <ProductID>1736</ProductID>
        <Version>
          <VersionMajor>4</VersionMajor>
          <VersionMinor>20</VersionMinor>
          <BuildMajor>0</BuildMajor>
          <BuildMinor>0</BuildMinor>
        </Version>
      </Creator>
  </Activity> """;

  final String tail = """    <Author xsi:type="Application_t">
    <Name>Connect Api</Name>
    <Build>
      <Version>
        <VersionMajor>0</VersionMajor>
        <VersionMinor>0</VersionMinor>
        <BuildMajor>0</BuildMajor>
        <BuildMinor>0</BuildMinor>
      </Version>
    </Build>
    <LangID>en</LangID>
    <PartNumber>006-D2449-00</PartNumber>
  </Author>
  </TrainingCenterDatabase>""";

  String activityBiking = """<Activity Sport="Biking">\n""";

  String activitiesContent = '';

  // Add Activity
  //-------------
  String activityContent = activityBiking;

  // Add ID
  activityContent = activityContent +
      addElement('Id', createTimestamp(tcxInfos.dateActivity));

  displayInfo(' $activityContent');

  // Add lap
  //---------
  String lapContent = '';
  lapContent = lapContent + addElement('TotalTimeSeconds', '1688.0');
  // Add Total distace in meters
  lapContent = lapContent + addElement('DistanceMeters', '88888.0');
  // Add Maximum speed in meter/second
  lapContent = lapContent + addElement('MaximumSpeed', '12.8888');
  // Add calories
  lapContent = lapContent + addElement('Calories', '12.8888');
  // Add intensity (what is the meaning?)
  lapContent = lapContent + addElement('Intensity', 'Active');
  // Add intensity (what is the meaning?)
  lapContent = lapContent + addElement('TriggerMethod', 'Manual');

  // Add track inside the lap
  String trackContent = '';
  int counterTrackpoint = 0;

  for (var point in tcxInfos.points) {
    String trackPoint = addTrackPoint(point);
    counterTrackpoint++;

    // To display the first 3 trackPoints
    if (counterTrackpoint < 4) {
      displayInfo(' $trackPoint');
    } // temp disp for the for loop

    trackContent = trackContent + trackPoint;
  }
  lapContent = lapContent + addElement('Track', trackContent);

  activityContent = activityContent +
      addAttribute('Lap', 'StartTime', createTimestamp(tcxInfos.dateActivity),
          lapContent);

  activityContent = activityContent + tailActivity;

  activitiesContent = addElement('Activities', activityContent);

  // Create the complete tcx file
  contents = prolog + activitiesContent + tail;

  sink.write(contents);
  // Close the file
  await sink.flush();
  await sink.close();

// return contents;
}

/// Generate a string that will include
/// all the tags corresponding to TCX trackpoint
///
/// Extension handling is missing for the moment
///
String addTrackPoint(TrackPoint point) {
  String _returnString;

  _returnString = "<Trackpoint>\n";
  _returnString = _returnString + addElement('Time', point.timeStamp);
  _returnString = _returnString +
      addPosition(point.latitude.toString(), point.longitude.toString());
  _returnString =
      _returnString + addElement('AltitudeMeters', point.altitude.toString());
  _returnString =
      _returnString + addElement('DistanceMeters', point.distance.toString());

  if (point.speed != null) {
    _returnString = _returnString + addExtension('Speed', point.speed);
  }

  if (point.power != null) {
    _returnString = _returnString + addExtension('Watts', point.power);
  }

  if (point.heartRate != null) {
    _returnString = _returnString + addHeartRate(point.heartRate);
  }

  _returnString = _returnString + "</Trackpoint>\n";

  return _returnString;
}

/// Add an extension like
///
///  <Extensions>
///              <ns3:TPX>
///                <ns3:Speed>1.996999979019165</ns3:Speed>
///              </ns3:TPX>
///            </Extensions>
///
/// Does not handle mutiple values like
/// Speed AND Watts in the same extension
///
String addExtension(String tag, double value) {
  String returnString;
  String extensionBeg = """<Extensions>\n   <ns3:TPX>\n""";
  String extensionMid;

  String extensionEnd = """   </ns3:TPX>\n</Extensions>\n""";

  double _value = value ?? 0.0;

  extensionMid =
      '     <ns3:' + tag + '>' + _value.toString() + '</ns3:' + tag + '>\n';

  returnString = extensionBeg + extensionMid + extensionEnd;

  return returnString;
}

/// Add heartRate in TCX file to look like
///
///       <HeartRateBpm>
///         <Value>61</Value>
///       </HeartRateBpm>
///
String addHeartRate(int heartRate) {
  String heartRateContentBeg = """
                 <HeartRateBpm>
              <Value>""";

  String heartRateContentEnd = """</Value>
            </HeartRateBpm>\n""";
  int _heartRate = heartRate ?? 0;
  String _valueString = _heartRate.toString();
  return heartRateContentBeg + _valueString + heartRateContentEnd;
}

/// create a position something like
/// <Position>
///   <LatitudeDegrees>43.14029800705612</LatitudeDegrees>
///   <LongitudeDegrees>5.771340150386095</LongitudeDegrees>
/// </Position>
String addPosition(String latitude, String longitude) {
  String returnString;
  returnString = '<Position>\n';

  returnString =
      returnString + '   <LatitudeDegrees>' + latitude + '</LatitudeDegrees>\n';
  returnString = returnString +
      '   <LongitudeDegrees>' +
      longitude +
      '</LongitudeDegrees>\n';

  returnString = returnString + '</Position>\n';

  return returnString;
}

/// create XML element
/// from content string
String addElement(String tag, String content) {
  String returnString;

  returnString = '<' + tag + '>' + content + '</' + tag + '>\n';

  return returnString;
}

/// create XML attribute
/// from content string

String addAttribute(
    String tag, String attribute, String value, String content) {
  String returnString;

  returnString = '<' + tag + ' ' + attribute + '="' + value + '">\n';
  returnString = returnString + content + '</' + tag + '>\n';

  return returnString;
}

/// Create timestamp for <Time> element in TCX file
///
/// To get 2019-03-03T11:43:46.000Z
/// utc time
/// Need to add T in the middle
String createTimestamp(DateTime dateTime) {
  String _returnString;

  _returnString = dateTime.toUtc().toString();
  _returnString = _returnString.replaceFirst(' ', 'T');

  return _returnString;
}


lib/rw_tcx.dart
library rw_tcx;

import 'dart:io';
import 'package:path_provider/path_provider.dart';

import 'models/TCXModel.dart';
import 'rTCX.dart';
import 'wTCX.dart';

Future<File>  _localFile(String fileName) async {
  final directory = await getApplicationDocumentsDirectory();
  var path = directory.path;
  return File('$path/fileName');
}

/// Generate the ride data structure
/// from a string
/// Not tested in detail because main goal
/// is to use writeTCX
///
TCXModel readTCX(String contents) {
  TCXModel rideData = TCXModel();
  rideData.points = List<TrackPoint>();

  // Search for the content of a tag
  Tag result = searchElement('Id', contents);
  print('result ${result.content}');


  Tag activity = searchAttribute('Activity', contents);
  print('Activity ${activity.content}');

  Tag totalDistance = searchElement('DistanceMeters', contents);
  print('Distance ${totalDistance.content}');

  Tag calories = searchElement('Calories', contents);
  print('Calories ${calories.content}');

  String speed = searchExtension('Speed', contents, index: 0);
  print('Speed $speed');

  // Start to store data in rideData structure
  rideData.totalDistance = double.tryParse(totalDistance.content);

  // search for sport= in the result and remove it
  List<String> _activity = activity.content.split('=');
  String activityType = _activity[1]; // Remove additional " "
  activityType = activityType.substring(1, activityType.length - 1);

  rideData.activityType = activityType;

  rideData.maxSpeed = double.tryParse(speed);
  rideData.calories = int.tryParse(calories.content);

  // Read now all the trackpoints
  //------------------------------
  bool noMoreTrackPoints = false;
  int idx = 0;

  

  do {
    TrackPoint trackPoint = extractTrackpoint(contents, idx);

    // Move to the next
    idx = trackPoint.index;
    print(
        'idx : ${trackPoint.index}  -- ${trackPoint.timeStamp} -- ${trackPoint.distance}');
    if (trackPoint.index == null) {
      // No more trackpoint available in the String
      print('No more trackPoints!');
      noMoreTrackPoints = true;
    } else {
      // Add a new point in rideData

      // If position is missing skip this trackpoint
      // in garmin connect sometimes position is missing in trackpoint
      if ((trackPoint.latitude != null) &&
          (trackPoint.longitude != null) &&
          (trackPoint != null)) {
        rideData.points.add(trackPoint);
      } else {
        print('skip this position ${trackPoint.timeStamp}');
      }
    }
  } while (!noMoreTrackPoints);
  return rideData;
}

// Generate the TCX file
/// from a TCX Model
///
Future<void> writeTCX(TCXModel tcxInfos, String filename) async {
  // Now generate a new file from rideData
  var generatedTCXFile = await _localFile(filename);
  var sink = generatedTCXFile.openWrite(mode: FileMode.writeOnly);

  String contents = '';

  // Generate the prolog of the TCX file
  String prolog = """ <?xml version="1.0" encoding="UTF-8"?>
    <TrainingCenterDatabase
    xsi:schemaLocation="http://www.garmin.com/xmlschemas/TrainingCenterDatabase/v2 http://www.garmin.com/xmlschemas/TrainingCenterDatabasev2.xsd"
    xmlns:ns5="http://www.garmin.com/xmlschemas/ActivityGoals/v1"
    xmlns:ns3="http://www.garmin.com/xmlschemas/ActivityExtension/v2"
    xmlns:ns2="http://www.garmin.com/xmlschemas/UserProfile/v2"
    xmlns="http://www.garmin.com/xmlschemas/TrainingCenterDatabase/v2"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:ns4="http://www.garmin.com/xmlschemas/ProfileExtension/v1">""";



  String tail = """
  <Creator>
  <Name>rw_tcx</Name>
  </Creator>
  </Activity>
  </Activities>
  </TrainingCenterDatabase>
  """;

  sink.write('$prolog\n');

  contents = prolog;

  String activitiesContent = '';

  // Add Activity
  //-------------
  String activityContent = '';

  // TODO to replace with real data
  var date = DateTime(2019, 6, 11, 16, 50);

  // Add ID
  activityContent =
      activityContent + '\n' + addElement('Id', createTimestamp(date));

  // Add lap
  //---------
  String lapContent = '';
  lapContent = lapContent + addElement('TotalTimeSeconds', '1688.0');
  // Add Total distace in meters
  lapContent = lapContent + addElement('DistanceMeters', '88888.0');
  // Add Maximum speed in meter/second
  lapContent = lapContent + addElement('MaximumSpeed', '12.8888');
  // Add calories
  lapContent = lapContent + addElement('Calories', '12.8888');
  // Add intensity (what is the meaning?)
  lapContent = lapContent + addElement('Intensity', 'Active');
  // Add intensity (what is the meaning?)
  lapContent = lapContent + addElement('TriggerMethod', 'Manual');

  // Add track inside the lap
  String trackContent = '';
  int counterTrackpoint = 0;

  for (var point in tcxInfos.points) {
    String trackPoint = addTrackPoint(point);
    counterTrackpoint++;

    // TODO: Remove the temp shortcut
    // Only the first 3 trackpoints
    //==============================
    if (counterTrackpoint == 4) {
      break;
    } // temp exit for the for loop?

    trackContent = trackContent + trackPoint;
  }
  lapContent = lapContent + addElement('Track', trackContent);

  activityContent = activityContent +
      addAttribute('Lap', 'StartTime', createTimestamp(date), lapContent);

  activitiesContent = addElement('Activities', activityContent);

  contents = prolog + activitiesContent + tail;


  // sink = generatedTCXFile.openWrite(mode: FileMode.append);
  sink.write(contents);
  // Close the file
  await sink.flush();
  await sink.close();

// Add the tail of the TCX file

// Check what is the result of the file generation
  sink = generatedTCXFile.openWrite(mode: FileMode.append);

  contents = await generatedTCXFile.readAsString();
  print('The file is ${contents.length} bytes long.');
  print(' $contents');

  // Close the file
  await sink.flush();
  await sink.close();
}


example/lib/main.dart
import 'package:flutter/material.dart';

import 'package:flutter/services.dart' show rootBundle;

import 'secret.dart'; // Where Strava app secret and clientId stored

import 'package:rw_tcx/models/TCXModel.dart';
import 'package:rw_tcx/rTCX.dart';
import 'package:rw_tcx/wTCX.dart';

import 'package:path_provider/path_provider.dart';  // Needed for getApplications

import 'package:strava_flutter/strava.dart';
import 'package:strava_flutter/Models/fault.dart';



// Used by Google sign in 
import 'package:google_sign_in/google_sign_in.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:googleapis/drive/v3.dart' as drive;
import 'dart:io' as io;


import 'package:http/http.dart'
    show BaseRequest, Response, StreamedResponse;
import 'package:http/io_client.dart';
class GoogleHttpClient extends IOClient {
  Map<String, String> _headers;

  GoogleHttpClient(this._headers) : super();

  @override
  Future<StreamedResponse> send(BaseRequest request) =>
      super.send(request..headers.addAll(_headers));

  @override
  Future<Response> head(Object url, {Map<String, String> headers}) =>
      super.head(url, headers: headers..addAll(_headers));

}



main(List<String> args) async {
  print('Start of rw_tcx example');


  // First init of Strava API stuff
  // Do authentication with the right scope
  final strava = Strava(
      true, // To get display info in API
      secret);

  bool isAuthOk = false;

  isAuthOk = await strava.oauth(clientId, 'activity:write', secret, 'auto');

  print('---> Strava Authentication result: $isAuthOk');


  // Start Google Auth to try access Google Drive

  final FirebaseAuth _auth = FirebaseAuth.instance;
  final GoogleSignIn googleSignIn = GoogleSignIn(
    scopes: ['https://www.googleapis.com/auth/drive'],
  );

  final GoogleSignInAccount googleSignInAccount = await googleSignIn.signIn();
  final GoogleSignInAuthentication googleSignInAuthentication =
      await googleSignInAccount.authentication;


   final AuthCredential credential = GoogleAuthProvider.getCredential(
    accessToken: googleSignInAuthentication.accessToken,
    idToken: googleSignInAuthentication.idToken,
  );

  // final FirebaseUser user = await _auth.signInWithCredential(credential);
  AuthResult authResult = await _auth.signInWithCredential(credential);

  assert(!authResult.user.isAnonymous);

 
  assert(await authResult.user.getIdToken() != null);

  final FirebaseUser currentUser = await _auth.currentUser();
  assert(authResult.user.uid == currentUser.uid);

  print('signInWithGoogle succeeded: ${authResult.user}');

  var client = GoogleHttpClient(await googleSignInAccount.authHeaders);

  var driveApi = drive.DriveApi(client);



  // Get the list of file from to root directory 
  // of the google drive
  // useful to check that driveApi is working 

  var fileList = await driveApi.files.list();
  // Display for 10 file names
  fileList.files.forEach((file) => print('---${file.name})'));


  print('Read a TCX file');

  // Use the asset file to test without having to create internally a ride
  TCXModel rideData = TCXModel();

  String contents = '';

  try {
    // Read the test file sample1.tcx in assets

    contents = await rootBundle.loadString('assets/sample2.tcx');

    print('contents $contents');

    rideData = readTCX(contents);
  } catch (e) {
    print('error loading assets file:  $e');
  }

  print('Write a TCX file');

  TCXModel tcxInfos = TCXModel();
    
  tcxInfos.points = rideData.points;
  tcxInfos.swVersionMajor = 'SWMajor';
  tcxInfos.swVersionMinor = 'SWMinor';
  tcxInfos.partNumber = 'Part Number';

  tcxInfos.dateActivity = DateTime(2019, 8, 15, 18, 28);


  await writeTCX(tcxInfos, 'generatedSample.tcx');

  print('End of write TCX test');



  final directory = await getApplicationDocumentsDirectory();
 
  String path = directory.path;

  var destFile = drive.File.fromJson({
    'mimeType': 'text/plain',
    'name': 'test_rw_tcx.txt', 
  });
  

  io.File fileToUpload  = io.File('$path/generatedSample.tcx');
  drive.Media mediaContent = drive.Media(fileToUpload.openRead(), fileToUpload.lengthSync());


  var resultFile = await driveApi.files.create(destFile,
  uploadMedia: mediaContent);



  // Now use Strava API to upload the generatedTCX
  // Strava authentication has been done previously
  String dirUpload = (await getApplicationDocumentsDirectory()).path;

  Fault fault = await strava.uploadActivity(
      'Test rw_TCX', 'It is working!', '$dirUpload/generatedSample.tcx', 'tcx');

  print('This is the end!!');



}

lib/models/TCXModel.dart
// TCXModel.dart

/// for the moment lap is not handled
/// handle speed extension
class TCXModel {
  String activityType;
  double totalDistance; // Total distance in meters
  double totalTime; // in seconds
  double maxSpeed; // in m/s
  int calories;
  String intensity;
  DateTime dateActivity; // Date of the activity
  List<TrackPoint> points;

  // Related to device that generated the data
  String creator;
  String deviceName;
  String unitID;
  String productID;
  String versionMajor;
  String versionMinor;
  String buildMajor;
  String buildMinor;

  // Related to software used to generate the TCX file
  String author;
  String name;
  String swVersionMajor;
  String swVersionMinor;
  String buildVersionMajor;
  String buildVersionMinor;
  String langID;
  String partNumber;

  // to get points
  List<TrackPoint> get getPoints {
    List<TrackPoint> returnpoints = List<TrackPoint>();
    for (var point in points) {
      returnpoints.add(point);
    }

    return returnpoints;
  }
}

class TrackPoint {
  double latitude; // in degrees
  double longitude;
  String timeStamp;
  double altitude; // in meters
  double speed; // Inst speed in m/s
  double distance; // in meters
  DateTime date;

  double cadence; // Not handled yet
  double power;
  int heartRate;

  int index; // position of the trackpoint in the string
  // used when reading the TCX file
}

class Tag {
  Tag({this.content, this.index});

  String content;
  int index; // Position of the last character of content in the string
  // used to read a TCX file
}


lib/rTCX.dart
// rTCX.dart

import 'models/TCXModel.dart';
import 'logTool.dart';

/// Generate the ride data structure
/// from a string
/// Not tested in detail because main goal
/// is to use writeTCX
///
TCXModel readTCX(String contents) {
  TCXModel rideData = TCXModel();
  rideData.points = List<TrackPoint>();

  // Search for the content of a tag
  Tag result = searchElement('Id', contents);
  displayInfo('result ${result.content}');

  Tag activity = searchAttribute('Activity', contents);
  displayInfo('Activity ${activity.content}');

  Tag totalDistance = searchElement('DistanceMeters', contents);
  displayInfo('Distance ${totalDistance.content}');

  Tag calories = searchElement('Calories', contents);
  displayInfo('Calories ${calories.content}');

  String speed = searchExtension('Speed', contents, index: 0);
  displayInfo('Speed $speed');

  // Start to store data in rideData structure
  rideData.totalDistance = double.tryParse(totalDistance.content);

  // search for sport= in the result and remove it
  List<String> _activity = activity.content.split('=');
  String activityType = _activity[1]; // Remove additional " "
  activityType = activityType.substring(1, activityType.length - 1);

  rideData.activityType = activityType;

  rideData.maxSpeed = double.tryParse(speed);
  rideData.calories = int.tryParse(calories.content);

  // Read now all the trackpoints
  //------------------------------
  bool noMoreTrackPoints = false;
  int idx = 0;

  do {
    TrackPoint trackPoint = extractTrackpoint(contents, idx);

    // Move to the next
    idx = trackPoint.index;
    // displayInfo(
    // 'idx : ${trackPoint.index}  -- ${trackPoint.timeStamp} -- ${trackPoint.distance}');
    if (trackPoint.index == null) {
      // No more trackpoint available in the String
      displayInfo('No more trackPoints!');
      noMoreTrackPoints = true;
    } else {
      // Add a new point in rideData

      // If position is missing skip this trackpoint
      // in garmin connect sometimes position is missing in trackpoint
      if ((trackPoint.latitude != null) &&
          (trackPoint.longitude != null) &&
          (trackPoint != null)) {
        rideData.points.add(trackPoint);
      } else {
        displayInfo('skip this position ${trackPoint.timeStamp}');
      }
    }
  } while (!noMoreTrackPoints);
  return rideData;
}

/// Tools to read a TCX and store the data in
/// a proper structure
///
///
///
/// Search for the content of a 'double' tag
///
/// What is between <tag> and </tag>
///
/// return index 0 and content '' if search is not successful
///
/// if index is missing start to search at the beginning of the contents string
Tag searchElement(String tag, String contents, {int index}) {
  String _startTag = '<' + tag + '>';
  Tag _returnTag = Tag();
  int _index = index ?? 0;

  int _pos = contents.indexOf(_startTag, _index);
  if (_pos == -1) {
    // displayInfo('start Element not found $_startTag');
    _returnTag.content = '';
  } else {
    // Search for end
    String _endTag = '</' + tag + '>';
    int _startPos = _pos + _startTag.length;

    int _endPos = contents.indexOf(_endTag, _startPos);
    if (_endPos == -1) {
      // Problem end tag not found
      displayInfo('End Element not found $_endTag');
      _returnTag.content = '';
    } else {
      // Get what is between the start tag and end tag
      _returnTag.content = contents.substring(_startPos, _endPos);
      _returnTag.index = _endPos;
    }
  }
  return _returnTag;
}

/// Search for the content of a 'single' tag
///
/// What is between '<tag' and  '>'
///
/// if index is missing start to search at the beginning of the contents string
/// return the content and the index of the last character of the content found
Tag searchAttribute(String tag, String contents, {int index}) {
  String _startTag = '<' + tag;
  Tag _returnTag = Tag();
  _returnTag.index = 0;
  int _index = index ?? 0;

  int _pos = contents.indexOf(_startTag, _index);
  if (_pos == -1) {
    displayInfo('start Attribute not found $_startTag');
    _returnTag.content = '';
  } else {
    // Search for end
    String _endTag = '>';
    int _startPos = _pos + _startTag.length;

    int _endPos = contents.indexOf(_endTag, _startPos);
    if (_endPos == -1) {
      // Problem end tag not found
      displayInfo('end Attribute not found $_endTag');
      _returnTag.content = '';
    } else {
      // Get what is between the start tag and end tag
      _returnTag.content = contents.substring(_startPos, _endPos);
      _returnTag.index = _endPos;
    }
  }
  return _returnTag;
}

/// Search extension that is nested in
///
/// like
///       <Extensions>
///            <ns3:TPX>
///              <ns3:Speed>2.7809998989105225</ns3:Speed>
///            </ns3:TPX>
///       </Extensions>
///
String searchExtension(String tag, String contents, {int index}) {
  Tag _returnTag = Tag();
  int _index = index ?? 0;
  // Search for the content of extensions
  Tag _insideExt = searchElement('Extensions', contents, index: _index);
  if (_insideExt.content != '') {
    // search now for what is between <ns3:tag> and </ns3:tag>
    _returnTag = searchElement('ns3:' + tag, _insideExt.content);
  }
  return _returnTag.content;
}

/// Get data from a trackpoint
///
/// latitude, longitude
/// altitude
/// speed
/// Give also the index of the last character of the trackpoint content
///
///        <Trackpoint>
///          <Time>2019-03-03T07:08:22.000Z</Time>
///          <Position>
///            <LatitudeDegrees>43.16713904030621</LatitudeDegrees>
///            <LongitudeDegrees>6.042847512289882</LongitudeDegrees>
///          </Position>
///          <AltitudeMeters>79.80000305175781</AltitudeMeters>
///          <DistanceMeters>2.75</DistanceMeters>
///          <Extensions>
///            <ns3:TPX>
///              <ns3:Speed>2.7809998989105225</ns3:Speed>
///            </ns3:TPX>
///          </Extensions>
///        </Trackpoint>
TrackPoint extractTrackpoint(String contents, int index) {
  TrackPoint returnTrackPoint = TrackPoint();

  // Search for the first trackpoint in the string contents
  Tag trackPointTag = searchElement('Trackpoint', contents, index: index);
  // Get the timestamp
  Tag timeStampTag = searchElement('Time', contents, index: index) ?? '';

  // Convert the timestamp into DateTime
  try {
    DateTime dateTime = DateTime.parse(timeStampTag.content);
    returnTrackPoint.date = dateTime;
  } catch (e) {
    displayInfo('Error in dateformat $e -> ${timeStampTag.content}');
    // This error will happen if it is the end of the trackpoints
    return TrackPoint();
  }

  // Get latitude and longitude
  Tag position = searchElement('Position', trackPointTag.content, index: 0);
  Tag latitude = searchElement('LatitudeDegrees', position.content);
  Tag longitude = searchElement('LongitudeDegrees', position.content);
  Tag altitude = searchElement('AltitudeMeters', trackPointTag.content);
  Tag distance = searchElement('DistanceMeters', trackPointTag.content);
  Tag heartRateWithValue = searchElement('HeartRateBpm', contents);
  // now remove 'value' tag
  Tag heartRate = searchElement('Value', heartRateWithValue.content);

  String speed = searchExtension('Speed', contents, index: index);
  String watts = searchExtension('Watts', contents, index: index);

  returnTrackPoint.latitude = double.tryParse(latitude.content);
  returnTrackPoint.longitude = double.tryParse(longitude.content);
  returnTrackPoint.altitude = double.tryParse(altitude.content);
  returnTrackPoint.distance = double.tryParse(distance.content);
  returnTrackPoint.heartRate = int.tryParse((heartRate.content));
  returnTrackPoint.timeStamp = timeStampTag.content;
  returnTrackPoint.speed = double.tryParse(speed);
  returnTrackPoint.power = double.tryParse(watts);
  returnTrackPoint.index = trackPointTag.index;

  if (timeStampTag.content == '') {
    return TrackPoint();
  } else
    return returnTrackPoint;
}
